subprojects/model-core/src/main/java/org/gradle/api/internal/provider/CredentialsProviderFactory.java

```diff
@@ -16,178 +16,196 @@
+    @Inject
+    protected abstract GradleProperties getGradleProperties();
+
+    String getValidatedIdentity() {
+        String identity = getParameters().getIdentity().getOrNull();
         if (identity == null || identity.isEmpty() || !identity.chars().allMatch(Character::isLetterOrDigit)) {
             throw new IllegalArgumentException("Identity may contain only letters and digits, received: " + identity);
         }
+        return identity;
     }
 
-    private <T extends Credentials> Provider<T> evaluateAtConfigurationTime(Callable<T> provider) {
-        return new InterceptingProvider<>(provider);
+    void assertRequiredValuesPresent() {
+        String identity = getParameters().getIdentity().get();
+        if (!missingProperties.isEmpty()) {
+            TreeFormatter errorBuilder = new TreeFormatter();
+            errorBuilder.node("The following Gradle properties are missing for '").append(identity).append("' credentials");
+            errorBuilder.startChildren();
+            for (String missingProperty : missingProperties) {
+                errorBuilder.node(missingProperty);
+            }
+            errorBuilder.endChildren();
+            throw new MissingValueException(errorBuilder.toString());
+        }
     }
 
-    private class InterceptingProvider<T> extends DefaultProvider<T> {
+    @Nullable
+    String getProperty(String propertyName) {
+        return getGradleProperties().find(propertyName);
+    }
 
-        public InterceptingProvider(Callable<? extends T> value) {
-            super(value);
+    @Nullable
+    String getRequiredProperty(String propertyName) {
+        String propertyValue = getProperty(propertyName);
+        if (propertyValue == null) {
+            missingProperties.add(propertyName);
         }
+        return propertyValue;
+    }
+}
 
-        @Override
-        public ValueProducer getProducer() {
-            calculatePresence(ValueConsumer.IgnoreUnsafeRead);
-            return super.getProducer();
-        }
+abstract class PasswordCredentialsValueSource extends CredentialsValueSource<PasswordCredentials> {
 
-        @Override
-        public boolean calculatePresence(ValueConsumer consumer) {
-            try {
-                return super.calculatePresence(consumer);
-            } catch (MissingValueException e) {
-                missingProviderErrors.add(e.getMessage());
-                if (consumer == ValueConsumer.IgnoreUnsafeRead) {
-                    return false;
-                }
-                throw e;
-            }
-        }
+    public PasswordCredentialsValueSource() {
+    }
+
+    @Nullable
+    @Override
+    public PasswordCredentials obtain() {
+        String identity = getValidatedIdentity();
+
+        String usernameValue = getRequiredProperty(identity + "Username");
+        String passwordValue = getRequiredProperty(identity + "Password");
+        assertRequiredValuesPresent();
```

Comment:
This method throws and breaks the (private) contract of ValueSource which is then compensated by a custom Provider implementation above. Perhaps spec of this value source can be reused somehow in `CredentialsPrerequisite` which would do the validation and build the error message for missing values.